src/server.ts

```diff
@@ -46,153 +43,22 @@ const interval = setInterval(gameLoop, 1000 / game.config.tickRate);
-    io.to("lobby").emit(
-      "newplayer",
-      JSON.stringify(Object.values(game.players)),
-    );
-  });
-
-  socket.on("logout", (message) => {
-    const payload: { id: string } = JSON.parse(message);
-    const { id } = payload;
-    delete game.players[id];
-    delete game.sockets[id];
-    const response: { players: Player[]; logoutPlayerId: string } = {
-      players: Object.values(game.players),
-      logoutPlayerId: id,
-    };
-    io.to("lobby").emit("logoutplayer", JSON.stringify(response));
-    console.log("Player : " + id + " quit the lobby.");
-  });
-
-  socket.on("start", (message) => {
-    const { id }: { id: string } = JSON.parse(message);
-
-    const player = Object.values(game.players).find(
-      (player) => player.id === id,
-    );
-
-    if (!player) {
-      socket.emit("error", JSON.stringify(UNAUTHORIZED));
-      return;
-    }
-
-    const unityPlayer = Object.values(game.players).find(
-      (player) => player.type === "UNITY",
-    );
-
-    if (!unityPlayer) {
-      io.emit("error", JSON.stringify(UNITY_PLAYER_NOT_FOUND));
-      return;
-    }
-
-    const shuffledPlayers = game.webplayers.sort(() => 0.5 - Math.random());
-    const half = Math.ceil(shuffledPlayers.length / 2);
-
-    // Fill evilmans
-    shuffledPlayers.slice(0, half).forEach((protector) => {
-      const player = game.players[protector.id];
-      const socket = game.sockets[protector.id];
-      if (player) {
-        game.players[player.id] = { ...player, role: "Protector" };
-      }
-      socket?.join("protectors");
-    });
-
-    // Fill protectors
-    shuffledPlayers.slice(half).forEach((protector) => {
-      const player = game.players[protector.id];
-      const socket = game.sockets[protector.id];
-      if (player) {
-        game.players[player.id] = { ...player, role: "Evilman" };
-      }
-      socket?.join("evilmans");
-    });
-
-    game.state.status = "STARTING";
-    game.state.startTimer = 5;
-
-    io.to("lobby").emit(
-      "start",
-      JSON.stringify({
-        player,
-      }),
-    );
-  });
-
-  socket.on("disconnect", () => {
-    if (game.state.status === "LOBBY") {
-      return;
-    }
-
-    const id = Object.entries(game.sockets).find(
-      ([, playerSocket]) => socket.id === playerSocket?.id,
-    )?.[0];
-
-    if (!id) {
-      return;
-    }
-
-    const player = game.players[id];
-    delete game.players[id];
-    delete game.sockets[id];
-
-    io.to("lobby").emit(
-      "newplayer",
-      JSON.stringify(Object.values(game.players)),
-    );
-    console.log(`${player?.name} (${player?.type}) Client disconnected`);
-  });
+  socket.on("whoami", (msg) => whoamiHandler.handleMessage(msg));
+  socket.on("signup", (msg) => signupHandler.handleMessage(msg));
+  socket.on("logout", (msg) => logoutHandler.handleMessage(msg));
+  socket.on("start", (msg) => startHandler.handleMessage(msg));
+  socket.on("disconnect", (msg) => disconnectHandler.handleMessage(msg));
```

Comment:
Here, it is important ot not do the following :
```ts
socket.on("whoami", whoamiHandler.handleMessage);
```
Typescript won't alarm us but doing this will means that the `this` keyword will not refer to the `WhoamiHandler` but will refer to the `socket.on` callback.